Faster minimum/maximum/extrema

mikmoore opened the pull request:
This PR uses a new method to greatly accelerate `minimum`/`maximum`/`extrema` for `IEEEFloat` values. The concept is to convert floats to signed integers that respect the same ordering that `min`/`max` do for floats. Since integers can be min/max'd much more quickly than floats, this leads to a considerable speedup. This is a minor extension to the concept used by `isless` since #39090. I used the changes here to slightly accelerate that version of `isless` as well.

Note that the old versions had special-cased `_mapreduce_impl` for `min`/`max`. The reckless application of these led to very poor performance for integer `minimum`/`maximum`. This implementation eliminates those specializations, improving integer performance as well.

Benchmark:
```
using BenchmarkTools
a = randn(1000);
b = zeros(1000);
c = @. ifelse(rand() < 0.1, missing, a);
j = rand(Int,1000);
@btime minimum($a); # "typical" case
@btime minimum($b); # pathological case for old version
@btime extrema($a);
@btime minimum($c); # with missing
@btime minimum($j); # integer performance
@btime searchsortedlast($a, x) setup=(x=randn()); # isless performance
```

Before:
```
  812.222 ns (0 allocations: 0 bytes)
  1.620 μs (0 allocations: 0 bytes)
  4.629 μs (0 allocations: 0 bytes)
  1.220 μs (0 allocations: 0 bytes)
  262.315 ns (0 allocations: 0 bytes)
  14.729 ns (0 allocations: 0 bytes)
```
This PR:
```
  173.764 ns (0 allocations: 0 bytes)
  173.721 ns (0 allocations: 0 bytes)
  278.276 ns (0 allocations: 0 bytes)
  954.839 ns (0 allocations: 0 bytes)
  92.453 ns (0 allocations: 0 bytes)
  11.311 ns (0 allocations: 0 bytes)
```
~#43725:~ EDIT: never mind - it sounds like I didn't get that benchmark right

Results are similar for `maximum`.
This idea does not support values that may be `Missing` whereas the original specialization did. Interestingly, eliminating that specialization (none of the new code here is run with `Missing`) still led to a performance improvement.

For `Float64`, this PR makes `minimum`/`maximum` 4.7x faster (more in pathological case), `extrema` 16x faster, and `isless` 1.3x faster. Note that `sort` already has specializations for floats, which remains faster than the naive application of this updated `isless`. This also makes `minimum`/`maximum` 2.8x faster for `Int`.


The technical idea here is finished, but I could use some help on the architecture. The current version makes `_mapreduce(f,op,...)` check for whether a registered acceleration function exists for `op` for the type of `f` applied to the return type. It only takes effect if the `return_type` of `f` applied to the input is known to be a `IEEEFloat`. In the case that such a function exists, it wraps `f` in the accelerating transformation and applies an appropriate inverse transformation to the result. Even if the architectural idea is sound, I'm not sure I put the corresponding functions in reasonable places within the code.

mikmoore committed:
faster minimum/maximum for IEEEFloat

mikmoore committed:
faster isless

mikmoore committed:
only apply _makefast if type is concrete

mikmoore committed:
fast extrema

N5N3 commented:
I'm affraid you didn't set up https://github.com/JuliaLang/julia/pull/43725 successfully. As all-zero input should not be a pain there.

This solution is good. The biggest concern might be it treats `NaN`s as ordered number:
```julia
julia> using QNaNs

julia> minimum([qnan(2),qnan(1)]) |> qnan
1
julia> minimum([qnan(1),qnan(2)]) |> qnan
1
```
I not sure whether @tkf would be happy with this.

If the core-devs are happy with this change.
I think we'd better do this optimization at `mapreduce_impl`. This would helps solve the Currently breakage on our type-based `reduce_empty`.

Also, I think it would be good to also accelerate slow dimensional reduction. e.g. (`minimum(a, dims = (1, 3))`).
To achieve this, I think we'd better make `floatorder_min` return a `float` and define a new reduction function  to handle them.

mikmoore committed:
less ambiguity in accelerator definitions

mikmoore committed:
only apply accelerator if it exists

mikmoore commented:
Thanks for the feedback.

I imagined I might have had that benchmarking wrong for that other PR. I removed the comparison from my earlier post.

You are correct that this version of `minimum` would differ from a nonspecializing `reduce(min,...)` in its preference among multiple NaNs. This may be objectionable.

I've fixed the breaking-empty-collections issue, although it still looks brittle to me.

I agree that this should also apply to dimensional reductions. I'm not sure I understand what you mean by "make `floatorder_min` return a float". While it's totally possible to define a pairwise `min` function based on this method, using it in a reduction involves redundant transformations to the integer space and so carries a significant performance penalty (such that other methods tend to perform better). Can you elaborate a little? 
For these reductions, the representations are bit-compatible so I might imagine doing the dimensional reduction into a `reinterpret(Int64,...)` array, then applying the inverse-transforms in-place and reinterpreting as `Float64`. Although that approach feels brittle, too.

The perfect world would involve a float `min` function that would vectorize under `reduce` -- none of this specialization would be needed at all. Unfortunately, I don't know compilers well enough to win that fight. My tinkering has been unsuccessful.

nalimilan commented:
Radix sort uses a similar approach:
https://github.com/JuliaLang/julia/blob/bd8dbc388c7b89f68838ca554ed7ba91740cce75/base/sort.jl#L1434-L1444

Cc: @LilithHafner

LilithHafner commented on base/reduce.jl:
```diff
@@ -410,7 +415,13 @@ The default is `reduce_first(op, f(x))`.
-_mapreduce(f, op, A::AbstractArrayOrBroadcasted) = _mapreduce(f, op, IndexStyle(A), A)
+function _mapreduce(f, op, A::AbstractArrayOrBroadcasted)
+    fT = _return_type(f, Tuple{eltype(A)}) # determine type of f(A[i])
+    preproc = _makefast_preproc(op, fT) # fetch an accelerator, if available
+    ffast = preproc == identity ? f : preproc ∘ f # apply accelerator, if available
```
```suggestion
    preproc = _makefast_preproc(op, fT) # fetch an accelerator, if available
    ffast = preproc === identity ? f : preproc ∘ f # apply accelerator, if available
```
Also, is this actually faster than
```suggestion
    _makefast_preproc(op, fT) ∘ f # apply accelerator, if available
```
<details>
<summary> This is what makes me question that </summary>

```julia
julia> f(x) = x^2
f (generic function with 1 method)

julia> using BenchmarkTools

julia> @belapsed map(identity ∘ f, 1:100)
7.436847710330139e-8

julia> @belapsed map(f, 1:100)
7.386489361702129e-8

julia> @belapsed map(identity ∘ f, 1:100)
7.409924487594391e-8

julia> @belapsed map(f, 1:100)
7.311134676564157e-8

julia> @belapsed map(identity ∘ f, 1:100)
7.403821656050955e-8

julia> @belapsed map(f, 1:100)
7.459745762711865e-8

julia> 

```

</details>

mikmoore replied:
There is no speed difference. However, the composition `identity \circ f` was breaking all the `reduce_empty` definitions. The ternary here keeps those from breaking in many cases, but I do worry it will still fail in some.

mikmoore replied:
I got some wires crossed. I'm implementing the second suggested change as well.

LilithHafner commented on base/float.jl:
```diff
@@ -414,19 +414,19 @@ end
-    return ifelse(ix < zero(IntT), ix ⊻ typemax(IntT), ix)
-end
-
 @inline function isless(a::T, b::T) where T<:IEEEFloat
```
I have a hard time believing that we can do any better than LLVM/hardware level IEEE floating-point comparison in general. Is there an LLVM intrinsic we can benchmark against?

mikmoore replied:
I don't know what LLVM intrinsics are available. But there isn't any x86 instruction for what we want. x86 comparisons and vminsd/vmaxsd have the wrong semantics with regard to signed zeros and NaNs.

This change was primarily intended to the `flipifneg` operation for re-use in the fast `min`/`max`. This was the fastest version of about 5 different methods I tried.

LilithHafner replied:
Lack of hardware support is surprising, but I believe you.

LilithHafner commented on base/reduce.jl:
```diff
@@ -604,62 +615,32 @@ julia> prod(1:5; init = 1.0)
+#   1) flip the ordering of all values with a negative sign
+#   2) circularly-shift the values so that NaNs are together at the correct end of the spectrum
+# The following functions perform this transformation and reverse it.
+# Under this scheme, we place a total order on every value (including NaNs), although the ordering of NaNs is an implementation detail.
```
I believe that which NaN is returned is, unlike in sorting, not specified in the documentation. Returning the first makes sense if it is just as easy, though in this case, I expect it is not.

JeffreySarnoff replied:
The IEEE Floating Point Standards Committee has decided that the most significant bit of a NaN cannot be discerned; the "sign bit" of a NaN must not play any role nor reflect payload content.  I initiated that discussion hoping to provide MPFR a reason to fix some unexpected interaction inside the string/print logic that flipped the state of the most significant bit of any NaN.

To use NaNs in a nonstandard way is now more gravel than roadblock.

LilithHafner replied:
@JeffreySarnoff Could you provide a link to this decision?

LilithHafner replied:
From the IEEE-754 standard, we have

> For an operation with quiet NaN inputs, other than maximum and minimum operations [maxNum and minNum as defined in the standard favor numbers over NaNs, unlike the maximum and minimum functions defined in Julia], if a floating-point result is to be delivered the result shall be a quiet NaN which should be one of the input NaNs.

I read this as: we should return one of the NaN inputs, but it is unspecified which.

mikmoore replied:
This PR chooses `sNaN < qNaN < -qNaN < -sNaN` (i.e., which end we choose from depends on `min` vs `max`), with further nuance based on the payload. Thus, what value is returned among multiple NaNs is dictated by the sign, qbit, and payload.

I can't look it up right now, but I believe our current implementations of `min` and `max` choose among NaNs based on sign and input position and our current `minimum` and `maximum` choose purely based on input position. The `min`/`max` methods being discussed under #45532 would change `min`/`max` to use the NaN semantics of either native machine `+` or `-` instead.

So the primary hazard I see for this method (and our existing implementations) is that it doesn't distinguish primarily based on signalingness, nor does it quiet a sNaN (although by adopting `+`/`-` semantics, one of the #45532 would likely be future-proof to such a requirement). But I don't believe either of those are required for these functions and they certainly aren't satisfied by the current versions.

JeffreySarnoff replied:
You read it correctly.  There is no "preferred" method.  Selecting the NaN with the larger payload as the one that continues offers better logics.
I have never owned hardware that uses that method.   Early on, I examined the NaN selection logic and tested payload propagation.  Those  choices are burned into processor design years before we can see one work.

The decision was taken over several days at the start of February 2021 using 
the IEEE-756 LISTSERVE.  Here is [the first message](https://listserv.ieee.org/cgi-bin/wa?A2=STDS-754;d95d7d2d.21&S=) and you may click on Next to page through. The last entry is from Mike Cowlishaw on Thursday, 4 Feb 2021 17:50:49.

LilithHafner commented on base/reduce.jl:
```diff
@@ -604,62 +615,32 @@ julia> prod(1:5; init = 1.0)
+#   2) circularly-shift the values so that NaNs are together at the correct end of the spectrum
+# The following functions perform this transformation and reverse it.
+# Under this scheme, we place a total order on every value (including NaNs), although the ordering of NaNs is an implementation detail.
+for (T,S) in ((Float16, Int16), (Float32, Int32), (Float64, Int64))
```
In sorting we do not provide an optimization for 16 bit floating-point numbers. I'm not sure why not, but I believe it was @StefanKarpinski who originally wrote the optimizations for 32 and 64 bit floating points and left out Float16. I imagine we should either optimize everywhere or nowhere, but I'm not sure what the reasoning is for either.

LilithHafner replied:
Sorting has since come around to this view of things and supports Float16. I see no reason why this shouldn't too.

LilithHafner reviewed (commented):
It would be nice to re-use from Sort.jl, but I believe the specifications here are sufficiently different with respect to handling NaNs that I don't think re-use is viable.

mikmoore committed:
Apply suggestions from code review

Co-authored-by: Lilith Orion Hafner <60898866+LilithHafner@users.noreply.github.com>

mikmoore commented:
I'm increasingly unhappy with how I've bolted this together. Hijacking the `f` function to do the transformation seems a little fragile. While the implementation here isn't too complicated, I'm not too excited to write the dimensional version (for, eg, `minimum(...;dims=...)`). I'm open to other ideas if people can think of something more composable.

N5N3 commented:
@[mikmoore] f85099cfd859d8d68f388716bda5318435a75a77 is a very primitive extension for slow dimensional reduction.
I removed the acceleration for `foldl` branch, as I believe we should inplement a general fast `mapreduce_impl` for non-linear reduction.

mikmoore commented:
> @mikmoore [f85099c](https://github.com/JuliaLang/julia/commit/f85099cfd859d8d68f388716bda5318435a75a77) is a very primitive extension for slow dimensional reduction.

I like that approach much better! I'll incorporate it when I have some time.

mikmoore committed:
reorganize

mcabbott commented:
I didn't manage to build this branch, but doing `@eval Base ...` to load its new functions into master, I see the following problems:
```
julia> extrema([1,2,Inf])
(1.0, 3.999999999999999)

julia> extrema([1,2,NaN])
(NaN, -1.3482698511467371e308)
```
These don't occur with `maximum`. They do occur with Float32.

N5N3 commented:
How did you set it?
My local build has pulled in a further version of this PR. And it has no such problem.
```julia
julia> extrema([1,2,Inf])
(1.0, Inf)

julia> extrema([1,2,NaN])
(NaN, NaN)

julia> extrema([1,2,NaN32])
(NaN32, NaN32)

julia> extrema([1,2,Inf32])
(1.0f0, Inf32)
```

N5N3 commented:
@mikmoore would you mind me push https://github.com/JuliaLang/julia/commit/f85099cfd859d8d68f388716bda5318435a75a77 here if you have no time to finish this. Since triage said we dont care about which NaN to return, I think it would be good to finish this PR and see does arm platform really need it?

mikmoore commented:
Sorry. I did spend some time a bit ago incorporating the sort of reorganization you suggested. I pushed that just now but wouldn't be upset if you integrated parts or all of your changes instead.

I spent some more time on this over the holidays but have been stuck on `minimum!`/`maximum!`. It's been a week or two, but as I recall they're causing allocations and are not as fast as they should be -- probably something to do with not specializing function arguments but I've been unable to fix them so far. Until that gets fixed I don't think this can be finished.

Here is the set I've been using to test this:
```julia
using BenchmarkTools
a = randn(1000);
b = @. ifelse(a <= 0, zero(eltype(a)), a); # convert all negatives to +0.0
d = @. ifelse(rand() < 0.1, missing, a); # convert some values to missing
m = reshape(a,20,50);
r = Matrix{Float64}(undef,size(m,1),1); c = Matrix{Float64}(undef,1,size(m,2));
j = rand(Int,1000);
@btime minimum($a); # "typical" case
@btime minimum($b); # pathological case for old version
@btime minimum($m;dims=1);
@btime minimum($m;dims=2);
@btime minimum!($c,$m);
@btime minimum!($r,$m);
@btime map!(minimum,$c,eachcol($m)); # this appears to be faster than the previous but shouldn't be
@btime map!(minimum,$r,eachrow($m)); # this appears to be faster than the previous but shouldn't be
@btime extrema($a);
@btime minimum($d); # with missing
@btime minimum($j); # integer performance
@btime searchsortedlast($a, x) setup=(x=randn()); # isless performance
sum(Float64[]) # check that this still works for empty collections

```

N5N3 committed:
Move the optimization deeper and spread it.

It helps to avoid extra allocation (caused by non-specialization)

N5N3 committed:
Turn off pairwise reduction for some op.

The accuracy won't be better.

N5N3 commented:
> have been stuck on minimum!/maximum!.

I'm afraid your change didn't touch the `minimum!`/`maximum!`. As they call `mapreducedim!` directly.
And I think `mapreducedim!` wont like `_makefast_mapreduction` as there's no guarantee that the dest array has the same eltype as `f.(A)`.

N5N3 committed:
Merge branch 'master' into pr/45581

N5N3 commented:
With [413b739](https://github.com/JuliaLang/julia/pull/45581/commits/413b7392c62c95e96742f838872b26e4c3cbcc1e) I have
```julia
julia> @btime minimum!($c,$m);
  861.905 ns (0 allocations: 0 bytes)

julia> @btime minimum!($r,$m);
  450.000 ns (0 allocations: 0 bytes)

julia> @btime map!(minimum,$c,eachcol($m));
  1.220 μs (0 allocations: 0 bytes)

julia> @btime map!(minimum,$r,eachrow($m));
  915.789 ns (0 allocations: 0 bytes)
```
So I think the original concern get resolved.

LilithHafner commented on base/reduce.jl:
```diff
@@ -1379,3 +1328,65 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
 end
+
+function _mapped_eltype(f, A)
+    elT = eltype(A) # try this
+    if elT === Any
+        elT = @default_eltype(A) # try again
+    end
```
Does the branch have any advantage over this?
```suggestion
    elT = @default_eltype A
```

LilithHafner commented on base/reduce.jl:
```diff
@@ -1379,3 +1328,65 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+        elT = @default_eltype(A) # try again
+    end
+    fT = _return_type(f, Tuple{elT}) # determine type of f(A[i]), if possible
+    return isconcretetype(fT) ? fT : Any # always widen this to `Any` if non-concrete
```
Why?

N5N3 replied:
Just a strong hint that this function is not that reliable.
As the current optimization only works for concrete cases.

LilithHafner reviewed (commented):
Have you considered using dispatch and a lazy map instead of hooks? i.e.
```julia
mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, [...]) =
    mapreduce_impl(_mapped_eltype(f, A), f, op, A, [...])
mapreduce_impl(_, f, op, A::AbstractArrayOrBroadcasted, [...]) =
    # existing implementation
for op in (min, max)
    @eval mapreduce_impl(::IEEEFloat, f, typeof($op), A::AbstractArrayOrBroadcasted, [...]) =
        postprocess(mapreduce_impl(f, fast_op, Iterators.map(preprocess, A), [...]))
end

N5N3 committed:
Adopt suggestion.

Co-Authored-By: Lilith Orion Hafner <60898866+LilithHafner@users.noreply.github.com>

N5N3 commented:
`Iterators.map` would return a `Generator` and fallback to `mapfoldl`. I'm not sure we want to add a similar hook there.

LilithHafner commented:
I mistakenly assumed that `Iterators.map(::AbstractArray)::AbstractArray` 💔

LilithHafner commented:
What about
```julia
for op in (min, max)
    @eval mapreduce_impl(::IEEEFloat, f, typeof($op), A::AbstractArrayOrBroadcasted, [...]) =
        postprocess(mapreduce_impl(preprocess ∘ f, fast_op, A, [...]))
end
```

N5N3 commented:
@mikmoore tried this at `_mapreduce` level see 81a0ff5c4f37c71a813338c17c7f08fb87a404f2.
But this approach seems not work well with `mapreducedim!` (as we only want to activate the optimization for dim1 reduction.)
For me it make sense to make `mapreducedim!` and `mapreduce_impl` share a similar code style.

LilithHafner commented:
That makes sense. Thanks for humoring me. It's frustrating how complex the `mapreduce` implementation is.

mikmoore commented on base/reducedim.jl:
```diff
@@ -301,14 +301,17 @@ function _mapreducedim!(f, op, R::AbstractArray, A::AbstractArrayOrBroadcasted)
+            R[i1,IR] = post(r)
         end
     else
         @inbounds for IA in CartesianIndices(indsAt)
```
Shouldn't it be possible to apply the same series of transformations when reducing along non-first dimensions? Or am I missing where that has already been handled?

LilithHafner replied:
Yes, exactly. And all the optimizations here are `reinterpret`-able.

N5N3 replied:
I dont think non-dim1 cases would be accelerated a lot as we will have to add a `map!(pre, dest, dest)` berfore the kernal loop, and a `map!(post, dest, dest)` after it.
And non dim1 minimum/maximum is vectorlized quite well
```julia
julia> @btime minimum($A, dims = 1);
  65.400 μs (7 allocations: 4.33 KiB)

julia> @btime minimum($A, dims = 2);
  56.500 μs (7 allocations: 4.33 KiB)
```
On my PC it's faster than dim1 with optimizaiton.

N5N3 replied:
A simple bench
```julia
f(b, c) = begin
    pre, op, post = Base._makefast_reducution(min, Float64)
    Base.initarray!(b, identity, min, true, c)
    map!(pre, b, b)
    Base.mapreducedim!(pre, op, b, c);
    map!(post, b, b)
end

julia> @btime f($B, $A);
  58.500 μs (0 allocations: 0 bytes)

julia> @btime minimum!($B, $A);
  57.700 μs (0 allocations: 0 bytes)
```

dpinol commented:
Hi,
any chance this PR could be moved on?
thanks

mikmoore commented:
I'm increasingly un-enamored with this PR. However, I still think that the current specializations should still be removed (or at least limited only to `IEEEFloat`) and the `isless` improvement seems fine.

But the rest of it is a lot of code and complication (and maintenance) for some modest improvements. Further, I suspect that this would be a regression on AMD systems where the desired `min`/`max` semantics are available as native instructions. 

Currently, the `reduce` family of functions are just too complicated to make specializations like this palatable. I suspect that's why those of us that worked on this eventually burned out. And there's further work happening on those families (universal pairwise `reduce`, changes to `init` semantics) that may some adjustment here.

But feel free to pick this up and get it over the line yourself, if you like. It would be easier (almost done, probably?) if you abandoned dimensional reduction and merely aimed to improve the whole-collection case.
